Add a Preview Bake button for quick iteration with LightmapGI

Calinou opened the pull request:
- `master` version of https://github.com/godotengine/godot/pull/45237.

When using the Preview Bake button instead of Bake Lightmaps, low quality settings will be used instead of settings from the LightmapGI node. A node configuration warning will be displayed
besides the LightmapGI node to inform the user that the bake was done in preview mode. This preview bake warning is persisted to storage to make teamwork easier.

This helps iterate quicker on a scene's lighting since you don't need to wait as much time to see results.

On a simple test scene, Preview Bake is about 3 times faster than a "final" bake with the default LightmapGI settings (if denoising is not the bottleneck). This metric will vary depending on CPU and GPU speed, as the baking process is performed on the GPU but denoising is performed on the CPU.

The preview bake settings can be adjusted in the Project Settings.

This also prints a message with the effective quality settings before beginning the lightmapping process.

**Testing project:** [test_lightmap_preview_bake_4.x.zip](https://github.com/godotengine/godot/files/12322526/test_lightmap_preview_bake_4.x.zip)

- This closes https://github.com/godotengine/godot-proposals/issues/2029.

## Preview ~~Bake~~

![image](https://github.com/godotengine/godot/assets/180032/81bfb004-2e74-4bad-8380-fd9d4543da16)

### Preview bake result

> [!NOTE]
>
> Tangents are a bit broken on this scene, so there are artifacts that would usually not appear.

*4 seconds to bake.*

![Screenshot_20240118_203800 webp](https://github.com/godotengine/godot/assets/180032/b11a3cc4-7088-4eef-809e-04592338cc60)

### "Final" bake result

*16 seconds to bake.*

![Screenshot_20240118_203828 webp](https://github.com/godotengine/godot/assets/180032/c76deec5-f728-438d-8ef7-a112f193c949)

Calinou committed:
Add a Preview Bake button for quick iteration with LightmapGI

When using the Preview Bake button instead of Bake Lightmaps,
low quality settings will be used instead of settings from the
LightmapGI node. A node configuration warning will be displayed
besides the LightmapGI node to inform the user that the bake
was done in preview mode. This preview bake warning is persisted
to storage to make teamwork easier.

This helps iterate quicker on a scene's lighting since you don't
need to wait as much time to see results.

On a simple test scene, Preview Bake is about 3 times faster
than a "final" bake with the default LightmapGI settings
(if denoising is not the bottleneck). This metric will vary depending
on CPU and GPU speed, as the baking process is performed on the GPU
but denoising is performedon the CPU.

The preview bake settings can be adjusted in the Project Settings.

This also prints a message with the effective quality settings
before beginning the lightmapping process.

Zireael07 commented:
Note that as mentioned on the 3.x PR, you *can* use the preview bake in your projects if your computer is too weak to handle the full bake.

verypleasentusername commented:
is this implemented?

YuriSizov commented:
@verypleasentusername No, this PR is still very much open.

Calinou commented:
Rebased and tested again, it works as expected.

- Add support for texel scale multiplier, which defaults to `0.5` (half resolution on each axis for preview bakes). This speeds up bakes *a lot* without having to disable denoising or directional baking.
- Change the default maximum number of bounces for preview bakes to 2, which looks significantly closer to high bounce counts compared to 1. The reduced texel scale makes it more viable.
- Removed toggle for denoiser and directional, as denoising is now very fast and directional impacts visuals too much to be worth toggling for preview bakes.

YuriSizov commented on editor/plugins/lightmap_gi_editor_plugin.cpp:
```diff
@@ -166,19 +169,30 @@ void LightmapGIEditorPlugin::bake_func_end(uint64_t p_time_started) {
+	bake->set_tooltip_text(TTR("Bakes lightmaps with the settings specified in the LightmapGI node."));
 	bake->set_text(TTR("Bake Lightmaps"));
 	bake->hide();
 	bake->connect("pressed", Callable(this, "_bake"));
```
So what exactly happens when you change it to `callable_mp`? There is no reason to keep using strings and keep binding this method.

Calinou replied:
If I use `callable_mp()` for the method without `.bind(false)` (the parameter is optional), I get an error when the editor opens and clicking the button does nothing.

I suppose I could always use `.bind(false)` or make the argument required though.

YuriSizov replied:
I think you should just bind false.

YuriSizov commented on editor/plugins/lightmap_gi_editor_plugin.cpp:
```diff
@@ -166,19 +169,30 @@ void LightmapGIEditorPlugin::bake_func_end(uint64_t p_time_started) {
+
 	bake = memnew(Button);
 	bake->set_theme_type_variation("FlatButton");
 	// TODO: Rework this as a dedicated toolbar control so we can hook into theme changes and update it
 	// when the editor theme updates.
```
This comment should probably be moved before both buttons now, since it is relevant to both of them.

ARez2 commented:
Could this be merged?

Calinou commented:
> Could this be merged?

4.3 is currently in release freeze, so new features can only be merged in 4.4 at the earliest.

ARez2 commented:
Yep! That's what I am doing. Good to hear that this will be considered for 4.4 though!

badsectoracula commented:
I fixed the conflicts and tested this (with some quick perusal of the source code). It is very useful as it allows for quick previewing of lighting (though some way to bake everything in the map is also needed).

I put the fixes here as the conflicts are actually straightforward even if they did confuse git somewhat: https://github.com/badsectoracula/godot/tree/lightmapping-preview-bake

Calinou commented:
Thanks for rebasing this PR! I'll incorporate the changes shortly.

> (though some way to bake everything in the map is also needed).

See https://github.com/godotengine/godot-proposals/issues/10323.

Calinou commented:
Rebased and tested again (with @badsectoracula's help), it works as expected.

There was some discussion recently about making baking and saving lightmaps separate operations, so that the preview bake could bake lightmaps without saving them. While this would be slightly faster (especially when developing on a device with slow I/O), it would also prevent using the preview lightmaps when playtesting a project, which can be problematic for some use cases. Sometimes, you want to be able to run the project with some kind of baked lightmaps, even if you don't have the time to do a full lightmap bake.

badsectoracula commented:
> There was some discussion recently about making baking and saving lightmaps separate operations, so that the preview bake could bake lightmaps without saving them. While this would be slightly faster (especially when developing on a device with slow I/O), it would also prevent using the preview lightmaps when playtesting a project, which can be problematic for some use cases. Sometimes, you want to be able to run the project with some kind of baked lightmaps, even if you don't have the time to do a full lightmap bake.

The editor already autosaves any modified scenes when you press the play button so it could also autosave the lightmaps when you want to play. I think overall being able to playtest without saving would be a good idea as you may want to try out things without saving - and could probably be implemented via some sort of saving any unsaved modified resources to a temporary file and then passing mappings like "map all loads for `foo.tscn` to `/tmp/blahblah/foo.tscn`" or something like that to the launched game process.

But i think this is something for a different topic/proposal than this and could be implemented in the future regardless of having preview bakes or not (which i find very useful) as i don't see any reason to not save (automatically like now or as a separate step in a future change) the preview lightmaps for gameplay testing. At the end it is just different settings for the lightmapper and you may even want to distribute builds with preview lightmaps (e.g. testers who could be testing your gameplay but not necessarily the final graphics will probably still need to know where there are shadows but not care about light leaks or noisy lightmaps).

KoBeWi commented:
Needs another rebase .-.

KoBeWi commented on editor/scene/3d/lightmap_gi_editor_plugin.cpp:
```diff
@@ -175,15 +178,25 @@ void LightmapGIEditorPlugin::bake_func_end(uint64_t p_time_started) {
+	bake_preview = memnew(Button);
+	bake_preview->set_theme_type_variation(SceneStringName(FlatButton));
+	bake_preview->set_button_icon(EditorNode::get_singleton()->get_editor_theme()->get_icon(SNAME("BakePreview"), EditorStringName(EditorIcons)));
+	bake_preview->set_tooltip_text(TTR("Bakes lightmaps with low-quality settings for quick iteration.\nPreview bake quality can be changed in the Rendering > Lightmapping > Preview Bake section of the Project Settings."));
+	bake_preview->set_text(TTR("Preview Bake"));
```
```suggestion
	bake_preview->set_tooltip_text(TTRC("Bakes lightmaps with low-quality settings for quick iteration.\nPreview bake quality can be changed in the Rendering > Lightmapping > Preview Bake section of the Project Settings."));
	bake_preview->set_text(TTRC("Preview Bake"));
```

Calinou commented:
> and there is no way to cancel it (the Cancel button does nothing).

This is being tracked in https://github.com/godotengine/godot/issues/104392.

Calinou commented:
Rebased and tested again, it works as expected.

I've added the confirmation dialog mentioned above:

<img width="700" height="175" alt="image" src="https://github.com/user-attachments/assets/3429ca77-123d-41b7-a31b-75b4d1e66108" />

However, it sometimes won't display and will confirm straight ahead (it might show up for a frame or two, not sure). I'm not sure what's causing this. cc @KoBeWi
